app/services/idv/financials_step.rb

```diff
@@ -1,32 +1,44 @@
 module Idv
   class FinancialsStep < Step
-    def complete?
-      idv_form.finance_type && idv_session.financials_confirmation.try(:success?) ? true : false
+    def submit
+      if complete?
+        @success = true
+        idv_session.financials_confirmation = true
+        idv_session.params = idv_form.idv_params
+      else
+        @success = false
+        idv_session.financials_confirmation = false
+      end
+
+      result
     end
 
     private
 
-    def vendor_validate
-      result = vendor_validator.validate
-      idv_session.params = idv_form.idv_params if complete?
-      result
+    attr_reader :success
+
+    def complete?
+      form_valid? && vendor_validator.success?
     end
 
     def vendor_validator_class
       Idv::FinancialsValidator
     end
 
-    def analytics_event
-      Analytics::IDV_FINANCE_CONFIRMATION
-    end
-
     def vendor_errors
-      idv_session.financials_confirmation.try(:errors)
+      vendor_validator.errors if form_valid?
     end
 
     def vendor_params
       finance_type = idv_form.finance_type
       { finance_type => idv_form.idv_params[finance_type] }
     end
+
+    def result
```

Comment:
> Perhaps because it requires any other object who wants to deal with this one to know that we return a hash with success and errors keys.

I don't understand this argument. Could you elaborate please? Whether a method returns a hash or an object that responds to methods, wouldn't you need to read the Class to know its API? If you want to deal with `#submit`, and it returned an object that responds to `#success` and `#errors`, you would still have to know that in advance, no? How else would you know how to interact with it?

The API we've established and agreed to use up until now is for classes that process the submission of data to return a hash, so it should be familiar to receive a hash with `success` and `errors` keys, and potentially other keys as well. I've also started documenting this in our wiki: https://github.com/18F/identity-private/wiki/Rails-Controller-Design-in-identity-idp

> Thoughts on having success and errors as methods instead?

If we're now saying that it's easier to work with methods than hashes, then I would suggest we have the classes return an [OpenStruct](https://ruby-doc.org/stdlib-2.3.3/libdoc/ostruct/rdoc/OpenStruct.html), rather than creating a new `FormResponse` type of class because that would mean creating a separate class to encapsulate the response for each form/service object because some need to return more than `success` and `errors`.

In fact, I used OpenStruct (which automatically gives you methods) the first time I implemented this pattern, but decided to use a hash instead because `OpenStruct`s are expensive. I didn't think it was worth the extra overhead just to be able to do `result.success` instead of `result[:success]`.

From the Ruby docs:
> An OpenStruct utilizes Ruby’s method lookup structure to find and define the necessary methods for properties. This is accomplished through the method method_missing and define_method.

> This should be a consideration if there is a concern about the performance of the objects that are created, as there is much more overhead in the setting of these properties compared to using a Hash or a Struct.